services/src/main/java/org/keycloak/authentication/requiredactions/UpdateProfile.java

```diff
@@ -73,36 +72,34 @@ public void processAction(RequiredActionContext context) {
         String oldFirstName = user.getFirstName();
         String oldLastName = user.getLastName();
         String oldEmail = user.getEmail();
-        UserProfileValidationResult result = forUpdateProfile(user, formData, context.getSession()).validate();
-        final UserProfile updatedProfile = result.getProfile();
-        List<FormMessage> errors = Validation.getFormErrorsFromValidation(result);
+        UserProfileProvider provider = context.getSession().getProvider(UserProfileProvider.class);
+        UserProfile profile = provider.forUpdate(UserProfileContext.UPDATE_PROFILE, formData, user);
+
+        try {
+            // backward compatibility with old account console where attributes are not removed if missing
+            profile.update(false, (attributeName, userModel) -> {
+                if (attributeName.equals(UserModel.FIRST_NAME)) {
+                    event.detail(Details.PREVIOUS_FIRST_NAME, oldFirstName).detail(Details.UPDATED_FIRST_NAME, user.getFirstName());
+                }
+                if (attributeName.equals(UserModel.LAST_NAME)) {
+                    event.detail(Details.PREVIOUS_LAST_NAME, oldLastName).detail(Details.UPDATED_LAST_NAME, user.getLastName());
+                }
+                if (attributeName.equals(UserModel.EMAIL)) {
+                    user.setEmailVerified(false);
+                    event.detail(Details.PREVIOUS_EMAIL, oldEmail).detail(Details.UPDATED_EMAIL, user.getEmail());
+                }
+            });
+
+            context.success();
+        } catch (ValidationException pve) {
+            List<FormMessage> errors = Validation.getFormErrorsFromValidation(pve.getErrors());
 
-        if (errors != null && !errors.isEmpty()) {
             Response challenge = context.form()
                     .setErrors(errors)
```

Comment:
Once we integrate ValidationSPI we should implement form.setErrors() method which directly accepts ValidationException and automatically handles it (with i18n) so this code will be simplified a bit and things will be hidden in the form framework.